velox/connectors/hive/HiveDataSink.cpp

```diff
@@ -101,4 +277,45 @@ bool HiveInsertTableHandle::isInsertTable() const {
   return locationHandle_->tableType() == LocationHandle::TableType::kExisting;
 }
 
+PartitionOrdinalIdGenerator::PartitionOrdinalIdGenerator(RowTypePtr inputType)
+    : inputType_(std::move(inputType)) {}
+
+void PartitionOrdinalIdGenerator::next(
+    const std::vector<DecodedVector>& input,
+    const SelectivityVector& activeRows,
+    std::vector<uint32_t>& result) {
+  auto numPartitions = inputType_->size();
+
+  result.resize(activeRows.size());
+
+  activeRows.applyToSelected([&](vector_size_t row) INLINE_LAMBDA {
+    uint32_t ordinalId = partitionToIdMap_.size();
+    const auto iter =
+        partitionToIdMap_.emplace(getPartitionValue(input, row), ordinalId)
+            .first;
+
+    result[row] = iter->second;
+  });
+}
+
+PartitionOrdinalIdGenerator::PartitionValue
+PartitionOrdinalIdGenerator::getPartitionValue(
+    const std::vector<DecodedVector>& inputVectors,
+    vector_size_t row) {
+  auto numPartitions = inputType_->size();
+
+  PartitionValue partition;
+  partition.columnValues.reserve(numPartitions);
+
+  for (column_index_t i = 0; i < numPartitions; i++) {
+    partition.columnValues.push_back(VALUE_ID_TYPE_DISPATCH(
+        makePartitionColumnValue,
+        inputType_->childAt(i)->kind(),
+        inputVectors[i],
+        row));
+  }
```

Comment:
Hi @mbasmanova , in terms of performance, is it okay to iterate over the list of decoded partition vectors by row first, and in each row cast every column to the right partition type?

If I go with the approach like UniqueValues w/ F14FastSet and customized Hasher as we discussed, I seem to have do the above to put every row into a structure (here is struct PartitionOrdinalIdGenerator::PartitionValue in HiveDataSink.h) and use the structure as the Map key.